conversion/conversion.go

```diff
@@ -0,0 +1,789 @@
+		p.MaybeReport(atomic.LoadInt64(&rows))
+		return nil
+	}
+	writer := spanner.NewBatchWriter(config)
+	conv.SetDataMode()
+	conv.SetDataSink(
+		func(table string, cols []string, vals []interface{}) {
+			writer.AddRow(table, cols, vals)
+		})
+
+	err := dynamodb.ProcessData(conv, dyclient)
+	if err != nil {
+		return nil, err
+	}
+	writer.Flush()
+	return writer, nil
+}
+
+type IOStreams struct {
+	In, SeekableIn, Out *os.File
+	BytesRead           int64
+}
+
+func schemaFromDump(driver string, ioHelper *IOStreams) (*internal.Conv, error) {
+	f, n, err := getSeekable(ioHelper.In)
+	if err != nil {
+		printSeekError(driver, err, ioHelper.Out)
+		return nil, fmt.Errorf("can't get seekable input file")
+	}
+	ioHelper.SeekableIn = f
+	ioHelper.BytesRead = n
+	conv := internal.MakeConv()
+	p := internal.NewProgress(n, "Generating schema", internal.Verbose())
+	r := internal.NewReader(bufio.NewReader(f), p)
+	conv.SetSchemaMode() // Build schema and ignore data in dump.
+	conv.SetDataSink(nil)
+	err = ProcessDump(driver, conv, r)
+	if err != nil {
+		fmt.Fprintf(ioHelper.Out, "Failed to parse the data file: %v", err)
+		return nil, fmt.Errorf("failed to parse the data file")
+	}
+	p.Done()
+	return conv, nil
+}
+
+func dataFromDump(driver string, config spanner.BatchWriterConfig, ioHelper *IOStreams, client *sp.Client, conv *internal.Conv, dataOnly bool) (*spanner.BatchWriter, error) {
+	// TODO: refactor of the way we handle getSeekable
+	// to avoid the code duplication here
+	if !dataOnly {
+		_, err := ioHelper.SeekableIn.Seek(0, 0)
+		if err != nil {
+			fmt.Printf("\nCan't seek to start of file (preparation for second pass): %v\n", err)
+			return nil, fmt.Errorf("can't seek to start of file")
+		}
+	} else {
+		// Note: input file is kept seekable to plan for future
+		// changes in showing progress for data migration.
+		f, n, err := getSeekable(ioHelper.In)
+		if err != nil {
+			printSeekError(driver, err, ioHelper.Out)
+			return nil, fmt.Errorf("can't get seekable input file")
+		}
+		ioHelper.SeekableIn = f
+		ioHelper.BytesRead = n
+	}
+	totalRows := conv.Rows()
+
+	p := internal.NewProgress(totalRows, "Writing data to Spanner", internal.Verbose())
+	r := internal.NewReader(bufio.NewReader(ioHelper.SeekableIn), nil)
+	rows := int64(0)
+	config.Write = func(m []*sp.Mutation) error {
+		_, err := client.Apply(context.Background(), m)
+		if err != nil {
+			return err
+		}
+		atomic.AddInt64(&rows, int64(len(m)))
+		p.MaybeReport(atomic.LoadInt64(&rows))
+		return nil
+	}
+	writer := spanner.NewBatchWriter(config)
+	conv.SetDataMode() // Process data in dump; schema is unchanged.
+	conv.SetDataSink(
+		func(table string, cols []string, vals []interface{}) {
+			writer.AddRow(table, cols, vals)
+		})
+	ProcessDump(driver, conv, r)
+	writer.Flush()
+	p.Done()
+
+	return writer, nil
+}
+
+func Report(driver string, badWrites map[string]int64, BytesRead int64, banner string, conv *internal.Conv, reportFileName string, out *os.File) {
```

Comment:
All public functions should have a comment. How about:

// Report generates a report of schema and data conversion.